### ISSUE TITLE
Why is $:/StoryList saved at first?

### ISSUE THREAD

**Issue by danielo515** [Jermolene/TiddlyWiki5#1742]
Hello,

I don't understand why the first thing TW does is to save the $:/StoryList tiddler. This can cause problems in some sync scenarios where we try to save a tiddler that already exists on the server. From my point of view what we should do is, when any syncadaptor is available, first load the $:/StoryList tiddler and then, if necessary , save it.

Regards

**Comment by pmario** [comment]
It will always be necessary, if you click a link and open a tiddler, because that modifies this tiddler. 

>This can cause problems in some sync scenarios where we try to save a tiddler that already exists on the server. 

IMO that's a bug from the server, if it can't handle this behaviour, because, that's what servers are designed for.

**Comment by danielo515** [comment]
>  IMO that's a bug from the server, if it can't handle this behaviour, because, that's what servers are designed for

I don't think so. This tiddler is just another tiddler, and should be loaded before you can change it and save it back again. But what TW is doing is modifying this tiddler and trying to save it BEFORE fetching the last version from the server. This is not consistent to what happens to other tiddlers, that are fully loaded when required.

**Comment by Jermolene** [comment]
>  But what TW is doing is modifying this tiddler and trying to save it BEFORE fetching the last version from the server.

That's by design; the most recent server version was baked into the HTML file. There's absolutely no requirement to read a tiddler before writing it.

**Comment by danielo515** [comment]
>  the most recent server version was baked into the HTML file.

Why are you supposing that?

> There's absolutely no requirement to read a tiddler before writing it.

It is a good idea to avoid "version collisions"

**Comment by pmario** [comment]
>> There's absolutely no requirement to read a tiddler before writing it.

>It is a good idea to avoid "version collisions"

@Jermolene In couchdb, there is the need to load every tiddler from the server, because they throw errors if try to overwrite an existing tiddler, with different or no internal ID. couchdb is a multi master db and they use this mechanism to avoid collisions.

**Comment by Jermolene** [comment]
> Why are you supposing that?

I'm explaining the reasoning behind the current design.

> @Jermolene In couchdb, there is the need to load every tiddler from the server, because they throw errors if try to overwrite an existing tiddler, with different or no internal ID. couchdb is a multi master db and they use this mechanism to avoid collisions.

Well, I'm still not convinced that that should change the contract between a sync adaptor and the sync mechanism. Surely the sync adaptor in such a case should recognise that it needs to take extra steps before writing the tiddler?

**Comment by danielo515** [comment]
Hello @Jermolene 

Is not so different if you think about it. What we are doing currently is asking for skinny tiddlers on startup. Then, why not checking those tiddlers and fetch those that are newer (higher revision) than the ones do we have before saving or editing them? If you save a tiddler that already exists on the server without any chick that will drive us into several sync problems. This will happen on any sync scenario, no matter which technology, fortunately some of them protects from this kind of problems. If we really want to provide sync capabilities, and we want to be able to provide a real multiuser experience we have to take this kind of things in account. Otherwise we will stay forever into a "personal, offline,isolated wiki"

**Comment by Jermolene** [comment]
Hi @danielo515

>  If you save a tiddler that already exists on the server without any chick that will drive us into several sync problems

That is what etags are for. We don't implement them properly yet, but they provide a robust way of detecting intervening changes on the server.

>  If we really want to provide sync capabilities, and we want to be able to provide a real multiuser experience we have to take this kind of things in account. Otherwise we will stay forever into a "personal, offline,isolated wiki"

As I've said elsewhere, I think the real issue with multiuser situations is not having a way to maintain per-user state on the server.
